### PR TITLE
Eager opgen support for in-place operations with variadic args

### PR DESCRIPTION
**Pull request by juanpaez22**
**Description**:
- Modified opgen for custom ops to use TORCH_LIBRARY binding frontend provided by PyTorch. This allows us to support additional bindings, e.g., TensorList, which does not bind properly using the PyBind interface.
- Extended opgen to support variadic in-place arguments and allow "void" return type. A simple for loop is used to copy input OrtValues into the output vector when the parameter is a TensorList or ArrayRef.
- Added test using existing Concat implementation as a custom op, which uses variadic inputs through TensorList.
- Modified existing custom op headers to better match ATen declarations.

**Motivation and Context**
- Custom ops may want to use TensorList as an input argument to allow variadic input tensors. TensorList binds to a list of tensors in Python.
- PyBind11 binding interface does not support binding to TensorList (throws TypeError- see https://github.com/pytorch/pytorch/issues/80979), but TORCH_LIBRARY interface does.
- In-place modification of variadic TensorList params is needed to support some ops (e.g., foreach_add)

### CODE REVIEW

File: orttraining/orttraining/eager/opgen/CustomOpDeclarations.h [hunk 1]
```diff
@@ -1,2 +1,3 @@
-Tensor gemm(const Tensor& A, const Tensor& B, const Tensor& C, float alpha, float beta, int transA, int transB);
-std::tuple<Tensor&, Tensor&, Tensor&> batchnorm_inplace(Tensor& X, const Tensor& scale, const Tensor& B, Tensor& input_mean, Tensor& input_var, const float epsilon, const float momentum); // {"schema": "batchnorm_inplace(Tensor(a!) X, Tensor scale, Tensor b, Tensor(b!) input_mean, Tensor(c!) input_var, float epsilon, float momentum) -> (Tensor(a!), Tensor(b!), Tensor(c!))", "dispatch": "False", "default": "True"}
\ No newline at end of file
+Tensor gemm(const Tensor& A, const Tensor& B, const Tensor& C, double alpha, double beta, int64_t transA, int64_t transB);
+std::tuple<Tensor&, Tensor&, Tensor&> batchnorm_inplace(Tensor& X, const Tensor& scale, const Tensor& B, Tensor& input_mean, Tensor& input_var, const double epsilon, const double momentum); // {"schema": "batchnorm_inplace(Tensor(a!) X, Tensor scale, Tensor b, Tensor(b!) input_mean, Tensor(c!) input_var, float epsilon, float momentum) -> (Tensor(a!), Tensor(b!), Tensor(c!))", "dispatch": "False", "default": "True"}
```
[the remark this answers is not in the document]
Comment (juanpaez22) [reply orttraining/orttraining/eager/opgen/CustomOpDeclarations.h:2]: Not that I know of. All aten ops in RegistrationDeclarations.h that use in-place TensorList return void. The other option is using std::vector<Tensor>, but this requires copying back. @souptc and I discussed this a bit ago, and I think we found that void return type is the most consistent with how aten ops are used today.

File: orttraining/orttraining/eager/opgen/opgen/generator.py [hunk 2]
```diff
@@ -439,24 +439,25 @@ def _write_function_body(self, writer: writer.SourceWriter, mapped_func: MappedO
...
+                            elif isinstance(return_info, ast.ArrayType):
+                                if self._is_inplace(return_info, torch_p):
+                                    writer.writeline(f"for (int i = 0; i < {onnx_op.outputs.count}; i++) {{")
+                                    writer.push_indent()
+                                    writer.writeline(
+                                        f"{onnx_op.outputs}[i] = ort_input_{onnx_op_index}_{onnx_op.inputs[input_index]}[i];"
+                                    )
+                                    writer.pop_indent()
+                                    writer.writeline("}")
+                                    in_place_params[0] = cpp_param.identifier.value
```
[the remark this answers is not in the document]
Comment (ashari4) [reply orttraining/orttraining/eager/opgen/opgen/generator.py:457]: i misread the code. in the in place case , the function returns void.

**Review by ashari4 — APPROVED** [review]

File: orttraining/orttraining/eager/test/ort_ops.py [hunk 3]
```diff
@@ -148,11 +148,18 @@ def test_batchnormalization_inplace(self):
         bias = torch.Tensor([0.0, 1.0]).to(device)
         mean = torch.Tensor([0.0, 3.0]).to(device)
         var = torch.Tensor([1.0, 1.5]).to(device)
-        y, mean_out, var_out = torch_ort.custom_ops.batchnorm_inplace(x, s, bias, mean, var, 1e-5, 0.9)
+        y, mean_out, var_out = torch.ops.ort.batchnorm_inplace(x, s, bias, mean, var, 1e-5, 0.9)
         assert torch.allclose(x.cpu(), y.cpu()), "x != y"
         assert torch.allclose(mean.cpu(), mean_out.cpu()), "mean != mean_out"
         assert torch.allclose(var.cpu(), var_out.cpu()), "var != var_out"
 
+    def test_variadic_inputs(self):
```
[the remark this answers is not in the document]
Comment (juanpaez22) [reply orttraining/orttraining/eager/test/ort_ops.py:156]: Doesn't seem like any kernels today use in-place and variadic args. We can add a test once we have such a kernel (e.g., element-wise operations like foreach_add).

**Pull request merged**